### PR TITLE
Added StartsWithTransfer and a test case

### PR DESCRIPTION
**Pull request by siddharthbe**

### CODE REVIEW

**Review by kelloggm — CHANGES_REQUESTED** [review]
The code looks fine. Can you add tests for the following, though, just to make sure they work as expected:
* a case where `String#startsWith` is called on a String with type `@StartsWithUnknown`
* a case where `String#startsWith` is called on a String with "https" as an argument, and then the string is assigned in the `else` block to a local variable with type `@StartsWith("https")`
* a `while` loop guarded by an `@StartsWith` annotation
* a test that confirms that this code works when the call to `startsWith` is within a boolean AND (`&&`) or a boolean OR (`||`). For example, test that a given String starts with either "https" or "jar", and confirm that its type inside the then block is `@StartsWith({"https", "jar"})`.

The only one of those I expect to fail is the last one. We can talk about how to (or whether) to support that tomorrow.

File: src/main/java/org/checkerframework/checker/startswith/StartsWithAnnotatedTypeFactory.java [hunk 1]
```diff
@@ -103,6 +103,10 @@ private void checkConcatenatedString (Tree tree, AnnotatedTypeMirror type) {
         }
     }
 
+    public AnnotationMirror getUNKNOWN() {
```
Comment (kelloggm) [inline src/main/java/org/checkerframework/checker/startswith/StartsWithAnnotatedTypeFactory.java]: It is considered bad style to use all caps in method names in Java. Can you change this to `getUnknown`? Or, better, give it a name that matches what it actually does: I usually prefer `getCanonicalTopAnnotation` or something like it.

Comment (kelloggm) [reply src/main/java/org/checkerframework/checker/startswith/StartsWithAnnotatedTypeFactory.java]: Also, is there any reason for this to be `public`? Package-private should be sufficient, right?

File: src/main/java/org/checkerframework/checker/startswith/StartsWithTransfer.java [hunk 2]
```diff
@@ -45,8 +52,37 @@ protected void processConditionalPostconditions(MethodInvocationNode node, Execu
...
         if (methodElement.equals(stringStartsWith)) {
             AnnotatedTypeMirror atm = aTypeFactory.getAnnotatedType(node.getArgument(0).getTree());
-            thenStore.insertValue(receiverReceiver, AnnotationUtils.getAnnotationByClass(atm.getAnnotations(),
-                                                                                         StartsWith.class));
+            thenStore.insertValue(receiverReceiver,
+                                  atm.getAnnotationInHierarchy(aTypeFactory.getUNKNOWN()));
+        }
+    }
+
+    /**For all then branches which have a (startsWith && startsWith) condition with the same receievers it annotates the
```
Comment (kelloggm) [inline src/main/java/org/checkerframework/checker/startswith/StartsWithTransfer.java]: Spelling of "receivers"

File: src/main/java/org/checkerframework/checker/startswith/StartsWithTransfer.java [hunk 3]
```diff
@@ -45,8 +52,37 @@ protected void processConditionalPostconditions(MethodInvocationNode node, Execu
...
+     * receiver with the GLB of the annotations of the arguments of startsWith
+     */
+    @Override
+    public TransferResult<CFValue, CFStore> visitConditionalAnd(ConditionalAndNode node,
+                                                                TransferInput<CFValue, CFStore> input){
+        TransferResult<CFValue, CFStore> result = super.visitConditionalAnd(node, input);
+        if(node.getLeftOperand().getTree().getKind() == Tree.Kind.METHOD_INVOCATION &&
+                node.getLeftOperand().getTree().getKind() == Tree.Kind.METHOD_INVOCATION){
```
Comment (kelloggm) [inline src/main/java/org/checkerframework/checker/startswith/StartsWithTransfer.java]: Shouldn't this be `getRightOperand`? If so, you should first add a test case that fails, and then fix it to be sure the test is fixed.

File: src/main/java/org/checkerframework/checker/startswith/StartsWithTransfer.java [hunk 4]
```diff
@@ -45,8 +52,37 @@ protected void processConditionalPostconditions(MethodInvocationNode node, Execu
...
+            MethodInvocationNode lft = (MethodInvocationNode) node.getLeftOperand();
+            MethodInvocationNode rht = (MethodInvocationNode) node.getRightOperand();
```
Comment (kelloggm) [inline src/main/java/org/checkerframework/checker/startswith/StartsWithTransfer.java]: What's the meaning of `lft` and `rht`? In Java, it's usually convention to give variables descriptive names - I would expect names like `leftType` and `rightType` or similar.

File: tests/url/TestTransfer.java [hunk 5]
```diff
@@ -1,21 +1,105 @@
...
+
+    void checkAndTwoReceiversFail(String s, String t) throws MalformedURLException{
+        if(s.startsWith("https://") && t.startsWith("http://")){
+            URL url = new URL(s);
+            // :: error: argument.type.incompatible
+            URL url2 = new URL(t);
+        }
+    }
+
+    void checkAndTwoReceivers(String s) throws MalformedURLException{
```
Comment (kelloggm) [inline tests/url/TestTransfer.java]: one receiver?

**Review by kelloggm — CHANGES_REQUESTED** [review]
Mostly minor changes, but I think I might have found a bug.

File: src/main/java/org/checkerframework/checker/startswith/StartsWithTransfer.java (hunk 3 above)
Comment (siddharthbe) [reply src/main/java/org/checkerframework/checker/startswith/StartsWithTransfer.java]: Yes, you're right. That was a bug in the program and I tested that by making a test case named "checkMethodInvocationRightOperandFail" which failed due to a cast exception and hence I fixed that method by calling getRightOperand() there.

File: src/main/java/org/checkerframework/checker/startswith/StartsWithAnnotatedTypeFactory.java [hunk 6]
```diff
@@ -103,6 +103,10 @@ private void checkConcatenatedString (Tree tree, AnnotatedTypeMirror type) {
         }
     }
 
+    AnnotationMirror getCanonicalTopAnnotation() {
```
Comment (kelloggm) [inline src/main/java/org/checkerframework/checker/startswith/StartsWithAnnotatedTypeFactory.java:109]: All methods should have a Javadoc comment explaining what they do.

File: src/main/java/org/checkerframework/checker/startswith/StartsWithTransfer.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        TransferResult<CFValue, CFStore> result = super.visitConditionalAnd(node, input);
+        if(node.getLeftOperand().getTree().getKind() == Tree.Kind.METHOD_INVOCATION &&
+                node.getRightOperand().getTree().getKind() == Tree.Kind.METHOD_INVOCATION){
+            MethodInvocationNode leftNode = (MethodInvocationNode) node.getLeftOperand();
+            MethodInvocationNode rightNode = (MethodInvocationNode) node.getRightOperand();
+            if(leftNode.getTarget().getReceiver().equals(rightNode.getTarget().getReceiver())) {
+                ExecutableElement methodElementLeft = leftNode.getTarget().getMethod();
+                ExecutableElement methodElementRight = rightNode.getTarget().getMethod();
+                if (methodElementLeft.equals(stringStartsWith) && methodElementRight.equals(stringStartsWith)) {
+                    Node receiver = leftNode.getTarget().getReceiver();
```
Comment (kelloggm) [inline src/main/java/org/checkerframework/checker/startswith/StartsWithTransfer.java]: You called `leftNode.getTarget().getReceiver()` already a few lines earlier. I suggest making it a local variable so you don't have to repeat yourself.

**Review by kelloggm — APPROVED** [review]
Looks good, modulo a few minor comments on style.

**Pull request merged**